File: backend/app/models/reservation.py
from sqlalchemy import Column, Integer, String, Text, Enum, DateTime, Date, Time, ForeignKey
from sqlalchemy.orm import relationship

from app.core.database import Base


class Reservation(Base):
    __tablename__ = "reservations"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    table_id = Column(Integer, ForeignKey("restaurant_tables.id"), nullable=True)
    reservation_code = Column(String(50), unique=True, nullable=False)
    reservation_date = Column(Date, nullable=False)
    start_time = Column(Time, nullable=False)
    end_time = Column(Time, nullable=False)
    number_of_guests = Column(Integer, nullable=False)
    customer_name = Column(String(100), nullable=False)
    customer_phone = Column(String(20), nullable=False)
    note = Column(Text, nullable=True)
    rejection_reason = Column(Text, nullable=True)
    status = Column(
        Enum("PENDING", "CONFIRMED", "CHECKED_IN", "CANCELLED", "COMPLETED", "NO_SHOW", "REJECTED"),
        nullable=False
    )
    created_at = Column(DateTime)
    updated_at = Column(DateTime)

    # Relationships
    user = relationship("User", back_populates="reservations")
    table = relationship("RestaurantTable", back_populates="reservations")
    table_session = relationship("TableSession", back_populates="reservation", uselist=False, cascade="all, delete-orphan")

File: frontend/src/types/reservation.ts
export type ReservationStatus = 
  | 'PENDING' 
  | 'CONFIRMED' 
  | 'CHECKED_IN' 
  | 'CANCELLED' 
  | 'COMPLETED' 
  | 'NO_SHOW' 
  | 'REJECTED';

export interface Reservation {
  id: number;
  user_id: number | null;
  table_id: number | null;
  reservation_code: string;
  reservation_date: string;
  start_time: string;
  end_time: string;
  number_of_guests: number;
  customer_name: string;
  customer_phone: string;
  note?: string;
  rejection_reason?: string;
  status: ReservationStatus;
  created_at: string;
  updated_at?: string;
}

export interface ReservationListResponse {
  items: Reservation[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface AssignTablePayload {
  table_id: number;
}

export interface RejectReservationPayload {
  reason: string;
}
